Custom error response handling

kellymclaughlin opened the pull request:
This is a change that stems from the discussion on #93. That PR was an attempt to fix a bug with returning a custom response body for requests for unmatched routes, but this PR represents a more broad effort to improve handling of requests for unmatched routes and also provide a means to return a custom response body for any error code. 

~~_Note:_ This work is currently incomplete. I have only added handling for the custom responses at `l07` in the decision tree. I'm opening the PR now to garner feedback on the approach before going any further.~~

Here is a short summary of these changes mostly taken from the commit message:
- Do not traverse the decision tree when no route matches. This mimics
  the [behavior of Erlang Webmachine](https://github.com/webmachine/webmachine/blob/master/src/webmachine_mochiweb.erl#L62-L64) and simplifies dealing with such
  requests. 
- Add the ability to specify a mapping of HTTP status codes to pairs of
  content type header values and custom response bodies for an Airship
  application or for an individual resource. The content type value in
  the mapping is used to set the response `Content-Type` header and the
  intention is that the specified content type value match the error
  response body.In the case of conflicts between application-level and
  resource-level error response mappings, the resource-level mappings
  are favored. This also more closely mimics what is offered by Erlang Webmachine.
- Replace the `Resource IO` parameter to `resourceToWai` with an
  ErrorResponses type that represents the mapping from HTTP status code
  to error response data. The 404 resource previously used to handle
  unmatched routes is no longer necessary since requests for unmatched
  routes now bail out early.

kellymclaughlin committed:
Custom error response handling

There are several changes represented in this commit related to custom
error responses in Airship.

* Do not traverse the decision tree when no route matches. This mimics
  the behavior of Erlang Webmachine and simplifies dealing with such
  requests.
* Add the ability to specify a mapping of HTTP status codes to lists of
  pairs of content type header values and custom response bodies for an
  Airship application or for an individual resource. Content negotiation
  is used to select the most appropriate error response from the
  provided list based on the client-provided Accept header value. The
  content type value in the mapping is used to set the response
  Content-Type header and the intention is that the specified content
  type value match the error response body. In the case of conflicts
  between application-level and resource-level error response mappings,
  the resource-level mappings are favored. The response is examined
  after the request is processed and if the response status is an error
  status and the response body is empty then a custom error response
  action may be invoked if a suitable match is found for the response
  status code and the Accept header value of the request.
* Replace the Resource IO parameter to resourceToWai with an
  ErrorResponses type that represents the mapping from HTTP status code
  to error response data. The 404 resource previously used to handle
  unmatched routes is no longer necessary since requests for unmatched
  routes now bail out early.

charleso commented:
@kellymclaughlin Despite my last, slightly confusing comment on your other PR this seems like a nice approach, and doesn't overload the `contentTypesProvided` field.

A few questions:
- The `type ErrorResponses = Map HTTP.Status (ByteString, ResponseBody)` seems a little restricted. Would it be expected to be able to handle different `Accept` types? Something like `Map HTTP.Status [NonEmpty (ByteString, ResponseBody)]` perhaps?
- I vaguely wonder whether you would want to be wrap the tuple in `m`? For example, would there be anything in the common `StateT s` that would be important/required in some cases?

Looks good though. :)

patrickt commented on src/Airship/Internal/Helpers.hs:
```diff
@@ -94,21 +95,40 @@ toWaiResponse Response{..} cfg trace quip =
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            (=<<) respond . run req $ do
```
This (the prefix-applied `bind`) is sort of an odd idiom here; I was able to decipher it after a while, but I'd avoid the one-liner here.

charleso replied:
> This (the prefix-applied bind) is sort of an odd idiom here;

Apologies, that was me, it's a bad habit of mine to avoid coming up with variable names.

patrickt commented on src/Airship/Internal/Route.hs:
```diff
@@ -80,14 +80,19 @@ star = Route [RestUnbound]
-route routes pInfo resource404 = foldr' (matchRoute pInfo) (resource404, (mempty, mempty)) routes
+route :: [(Route, a)]
+      -> [Text]
+      -> Maybe (a, (HashMap Text Text, [Text]))
```
When I see doubly-nested tuple types—note that we have an `(a, (b, [c]))` structure here, all wrapped in a Maybe—I find it hard to juggle the types in my head. Can we use a data structure for these various types, or, failing that, rub some `type` aliases on them? At the very least, we can destructure a couple of those tuples.

patrickt replied:
There also might be more expressive alternatives to folds we can think of to build routing primitives. `mapAccumL` is a good one. This could all be addressed in another PR though.

kellymclaughlin replied:
I've created a simple type alias, `MatchedRoute`, that reduces some of the line noise, but I'm afraid it doesn't do much to increase clarity. I think doing anything more, however, goes beyond the scope of this PR and is better handled separately.

patrickt commented on src/Airship/Resource.hs:
```diff
@@ -141,4 +142,5 @@ defaultResource = Resource { allowMissingPost          = return False
                            , serviceAvailable          = return True
                            , uriTooLong                = return False
                            , validContentHeaders       = return True
+                           , errorResponses            = empty
```
ultra-pedantry: use `mempty` here and you won't have to use that import line

patrickt commented on src/Airship/Types.hs:
```diff
@@ -101,19 +104,21 @@ data ResponseBody
 
 type Trace = [Text]
 
+type ErrorResponses = Map HTTP.Status (ByteString, ResponseBody)
```
Nitpick: I would actually use a singular noun for this type alias. `ErrorRecord`? `FailedResponseRecord`? Ehh. I'm failing to convince myself, really. Whatever.

patrickt replied:
Also can we use a `HashMap` here instead of a Map? Would kill our extra `containers` dependency, though not by much.

patrickt commented on src/Airship/Internal/Decision.hs:
```diff
@@ -564,7 +565,13 @@ l07 r = do
     if requestMethod req == HTTP.methodPost
         then m07 r
-        else lift $ halt HTTP.status404
+        else do
```
can you use one `lift` here instead of the three later on?

patrickt commented on src/Airship/Internal/Helpers.hs:
```diff
@@ -94,21 +95,40 @@ toWaiResponse Response{..} cfg trace quip =
+        Nothing -> do
+            -- No route matched so bail out
+            let (ct, respBody) = fromMaybe (mempty, Empty) response404
+                resp = Response HTTP.status404 [(HTTP.hContentType, ct)] respBody
```
weird pedantry, but can we add another `let` there? I know it hangs over from the previous `let`, but the indentation confused me.

patrickt commented on airship.cabal:
```diff
@@ -42,6 +42,7 @@ library
                       , bytestring
                       , bytestring-trie == 0.2.4.*
                       , case-insensitive
+                      , containers
```
see my later comments on wondering if we can get away with a HashMap

tmcgilchrist commented:
Nice work, I do think you'd want access to the state when generating the response for errors like @charleso said.

kellymclaughlin commented:
@charleso Thanks for those questions and feedback. For your first question, I had the handling the differen `Accept` types in my mind for this, but wasn't sure if that would be useful, but since you bring it up as well I think I will add that in. For the 2nd question about state access, it seems like a good idea so I'll gladly incorporate that.

Thanks also for the feedback @patrickt and @tmcgilchrist! I'll work on incorporating this feedback and completing the changes to the decision module and reping this PR when I have it finished.

kellymclaughlin commented:
Ok, I've completed a reworking of the original idea based on the previous feedback. The content-type of the error response for a given status code is now negotiated based on the request's Accept header value. The type of the error responses that users specify is now `Webmachine m ResponseBody`. As an example I modified the Basic example to illustrate accessing state and performing an IO action on a 404:

```
main :: IO ()
main = do
    static <- staticResource Cache "assets"
    let port = 3000
        host = "127.0.0.1"
        settings = setPort port (setHost host defaultSettings)
        routes = myRoutes static
    mvar <- newMVar HM.empty
    let s = State mvar
    putStrLn "Listening on port 3000"
    runSettings settings (resourceToWaiT defaultAirshipConfig (const $ flip evalStateT s) routes errors)
    where
        errors :: (MonadIO m, MonadState State m) => M.Map HTTP.Status [(MediaType, Webmachine m ResponseBody)]
        errors = let _response404 = escapedResponse "<html><head></head><body><h1>404 Not Found</h1></body></html>"
                 in M.singleton HTTP.status404 [("text/html", response404Action)]

response404Action :: (MonadIO m, MonadState State m) => Webmachine m ResponseBody
response404Action = do
    s <- lift get
    _m <- liftIO (readMVar (_getState s))
    liftIO $ print ("Hooray, IO!" :: String)
    return $ escapedResponse "<html><head></head><body><h1>404 Not Found</h1></body></html>"
```

It's a contrived example, but it shows that there could be a lot of useful possibilities for customizing an application or resource. I have updated all calls to `halt` for 4XX or 5XX in the Decision module to instead be calls to `haltWithErrorResponse`. 

Patrick had previously asked about using `HashMap` instead of `Map` to avoid bringing in the `containers` package. The reason I went with `Map` is that the `Network.HTTP.Types.Status` type does not have a `Hashable` instance so instead of going with just integers to represent the status codes I used `Map`.

charleso commented on src/Airship/Internal/Decision.hs:
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
+    req <- request
+    let reqHeaders = requestHeaders req
+        acceptStr = lookup HTTP.hAccept reqHeaders
+        errBodies = map dupContentType <$> M.lookup HTTP.status404 errorResponses
```
I'm assuming the `status404` is a bug here and should be `status`?

kellymclaughlin replied:
Good catch, definitely a bug. Will fix.

charleso commented on src/Airship/Internal/Decision.hs:
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
+        errBodies = map dupContentType <$> M.lookup HTTP.status404 errorResponses
+        mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+    when (isJust mResp) $ do
+       let (Just (ct, body)) = mResp
```
Would suggest pattern matching or `forM_` here rather than the `isJust` and unsafe  pattern matching

charleso commented on src/Airship/Types.hs:
```diff
@@ -217,6 +228,29 @@ eitherResponse reqDate reqParams dispatched req resource = do
+            (Just (c, b)) -> do
+                (c, b)
+            Nothing ->
+                ("text" // "plain", return Empty)
```
`fromMaybe ("text" // "plain", return Empty) mResp` /codegolf

charleso commented on src/Airship/Types.hs:
```diff
@@ -217,6 +228,29 @@ eitherResponse reqDate reqParams dispatched req resource = do
+                 UTCTime
+              -> Request
+              -> ErrorResponses m
+              -> m Response
```
I'm not sure but I suspect if you make this function return `Webmachine m Response` you can remove some of the duplication with `resourceToWaitT` (eg. `runRWST`) down below.

charleso commented on src/Airship/Internal/Decision.hs:
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
+-- | This is a shortcut for constructing a 'Response' with 'getResponseHeaders' and 'getResponseBody'
+-- | and passing that response to 'finishWith'.
+haltWithErrorResponse :: Monad m => HTTP.Status -> Resource m -> Webmachine m a
+haltWithErrorResponse status Resource{..} = do
```
This suggestion might be a pain.

One small problem with having something different to `halt` is that consumers will not get this functionally for free (if they're naughty and use `halt` that is), as well as you having to change all of the `Decision.hs` functions manually.

One suggestion which you should feel _completely_ free to ignore would be to do (something like) inspect the `Response` in `resourceToWaiT` and if no body is set call `errorResponse`. The danger I see is if you really really want to halt with a status but no body and also have an error handler for the same status. Just an idea, I'm curious if anyone has an opinion on that one.

kellymclaughlin replied:
Yes, I had a similar thought about whether or not this was a case that my be of interest/concern to users. I like idea of checking for a response body and only falling back to the resource's default error responses if it is not set. 

The case where someone has defined an error response for a particular status code and then wants to return a response with no body at some point in a resource seems to me like enough of an edge case that I can live with making the path to accomplish that to not define an error response for that status code for the resource and then the user must set the response body manually at the points in their resource code where they would like it returned. 

I'll do some playing around with this and see what comes of it.

charleso commented:
Looking good @kellymclaughlin

kellymclaughlin commented:
Reworked the approach is based on [feedback](https://github.com/helium/airship/pull/94#discussion-diff-56925630) from @charleso. This adjusted approach yields a much smaller set of changes that still achieves the same results.

tmcgilchrist commented on src/Airship/Types.hs:
```diff
@@ -140,6 +147,7 @@ instance MonadBaseControl b m => MonadBaseControl b (Webmachine m) where
 -- | A convenience synonym that writes the @Monad@ type constraint for you.
 type Handler m a = Monad m => Webmachine m a
```
Hmm, thought we got rid of this synonym?

reiddraper replied:
Looks like it snuck back in, in 3c97d750.

reiddraper replied:
Or rather, was never removed?

tmcgilchrist replied:
Might be remembering we removed it on our branch.

I don't think it should be there, but we have had that conversation before :-) . I'm curious if you're using it anywhere?

reiddraper replied:
I think we removed our use of it, too. Definitely happy to get rid of it.

tmcgilchrist replied:
https://github.com/helium/airship/pull/98

On Tue, Mar 29, 2016 at 9:13 AM, Reid Draper notifications@github.com
wrote:

> In src/Airship/Types.hs
> https://github.com/helium/airship/pull/94#discussion_r57642647:
> 
> > @@ -140,6 +147,7 @@ instance MonadBaseControl b m => MonadBaseControl b (Webmachine m) where
> >  -- | A convenience synonym that writes the @Monad@ type constraint for you.
> >  type Handler m a = Monad m => Webmachine m a
> 
> I think we removed our use of it, too. Definitely happy to get rid of it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/helium/airship/pull/94/files/f9ade37a41c7525f6be30ea2cfc32c71c5929578#r57642647

tmcgilchrist commented:
Less changes for the same behaviour is a win. :rocket: 
I find the `resourceToWai` and `errorResponse` functions more complicated than I'd like but that's just personal preference.

This is good by me :+1:

charleso commented on src/Airship/Internal/Helpers.hs:
```diff
@@ -94,21 +97,74 @@ toWaiResponse Response{..} cfg trace quip =
+               errBodies = map dupContentType <$> M.lookup _responseStatus errResps
+               mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+           when (isJust mResp) $ do
+               let (Just (ct, body)) = mResp
```
Just a nit reminder, but would definitely prefer to use `forM` here rather than having to use an unsafe cast.

``` haskell
forM_ mResp $ \(ct, body) -> do
  putResponseBody =<< body
  addResponseHeader ("Content-Type", renderHeader ct)
```

kellymclaughlin replied:
Thanks for the reminder! I just forgot about that bit of feedback. Will change.

charleso commented on src/Airship/Internal/Helpers.hs:
```diff
@@ -94,21 +98,73 @@ toWaiResponse Response{..} cfg trace quip =
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            respond =<< run req (do
```
Vaguely wondering if it's worth removing some of the dupe here? Something along the lines of. EDIT: Realised you don't need to update the `errorResponses` on the resource any more 

``` haskell
let (er, (params, matched), r) = case route routeMapping pInfo of
  Nothing ->
    (errors, (mempty, []), return $ Response HTTP.status404 [(HTTP.hContentType, "text/plain")] Empty)
  Just (resource, pm) ->
    (M.union (errorResponses resource) errors, pm, flow resource)
respond =<< run req (do
  (response, trace) <- eitherResponse nowTime params matched req (r >>= errorResponse er)
  return $ toWaiResponse response cfg (traceHeader trace) quip)
```

charleso commented:
:+1: Nice work Kelly.

charleso commented on src/Airship/Internal/Helpers.hs:
```diff
@@ -94,21 +98,73 @@ toWaiResponse Response{..} cfg trace quip =
+-- | If the Response body is Empty the response body is set based on the error responses
+-- | provided by the application and resource. If the response body is not Empty or
+-- | there are no error response configured for the status code in the Response then no
+-- | action is taken. The contents of the 'Webmachine'' response body will be streamed
```
Super nit, but the extra `|` on lines 2+ doesn't seem right. Not sure what haddock would do with them.
